File: web-server/app/service/UserService.java
package service;

import model.Advert;
import model.User;
import repository.interops.AdvertRepositoryJv;
import repository.interops.UserRepositoryJv;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class UserService {

    @Inject
    private AdvertRepositoryJv advertMapper;

    @Inject
    private UserRepositoryJv userMapper;

    public User getUserForAdvert(int advertId) {
        Advert advert = advertMapper.findById(advertId);
        if (advert == null) {
            return null;
        }
        return userMapper.getUserForAdvert(advert.id());
    }

}

File: web-server/app/service/UserService.scala
package service

import javax.inject.Inject

import doobie.imports._
import fs2.interop.cats._
import model.User
import repository.{AdvertRepository, DbConnection, UserRepository}

class UserService @Inject() (dbConnection: DbConnection, advertRepository: AdvertRepository, userRepository: UserRepository) {

  def getUserForAdvert(advertId: Int): Option[User] =
    userRepository.getUserForAdvert(advertId)
      .option
      .transact(dbConnection.xa)
      .unsafePerformIO

}
